archive/tar.go

```diff
@@ -131,6 +131,17 @@ func applyNaive(ctx context.Context, root string, tr *tar.Reader, options ApplyO
 		convertWhiteout = options.ConvertWhiteout
 	)
 
+	// flush data into disk
+	defer func() {
+		if err != nil {
+			return
+		}
+
+		if err = dirsync(root); err != nil {
```

Comment:
I am not sure that every filesystem supports dir sync, at least, the windows doesn't support it.

```
issues_test.go:53: failed to sync root C:\Users\appveyor\AppData\Local\Temp\1\prefix-test552602531: sync C:\Users\appveyor\AppData\Local\Temp\1\prefix-test552602531: The handle is invalid.
```

Reply:
This will force a sync of any other writer on that volume?

Reply:
will update

Reply:
I mostly want more evidence (or documentation) this actually does what it is intending, ensuring that everything under that directory is also synced. If it works on a subset of filesystem types, that is fine.

Reply:
I changed the behaviour because the `fsync` dir doesn't sync data into filesystem. As the `fsync` manpage mentioned, we need the sync fo regular file.

```
  Calling fsync() does not necessarily ensure that the entry in the directory containing the file has also reached disk.  For that an explicit fsync() on a file descriptor for the directory is also needed.
``` 

So I did sync for each regular file and `symlink/hardlink/chardevice...` are presented by inode in *nix which will be synced into filesystem. 

For windows, I didn't find any office document about syncing dir things. But I got some issue https://github.com/dgraph-io/badger/issues/699#issuecomment-504133587 which mentions that the windows doesn't support syncing dir into filesystem.